### PR TITLE
Auth for real, yo!

### PR DESCRIPTION
**Pull request by parndt**
This replaces #2908

### CODE REVIEW

**Comment by bricesanchez** [comment]
At this time, the only problem is the site_bar who is always dispalying, i'm trying to find a fix.

**Comment by parndt** [comment]
That's because the user is always logged in, because the default implementation has no log out.

**Comment by bricesanchez** [comment]
Yes, but when i use our Spree-Refinery connector, it should be the Spree::User and not the NilUser of refinery

**Comment by bricesanchez** [comment]
Write cache fails in production : https://github.com/refinery/refinerycms/blob/auth-for-real-yo-tmp/pages/app/controllers/refinery/pages_controller.rb#L88

**Comment by bricesanchez** [comment]
Could we merge master to this branch, in order to be up-to-date ?

**Comment by simi** [comment]
feel free to

**Comment by bricesanchez** [comment]
I've done the rebase to master and git push -f this branch.

I'd created a "backup branch" before : https://github.com/refinery/refinerycms/tree/auth-for-real-yo-tmp-backup-brice

**Comment by bricesanchez** [comment]
I've fixed `refinery_user?` errors for site_bar and write_cache? in our spree-refinery-authentication gem. 

I've no blocker to don't merge this PR :)

**Comment by bricesanchez** [comment]
...except a commit squash :P

**Comment by parndt** [comment]
Cool, the main blocker is that this will result in a non working authentication for anyone using it as we haven't converted refinerycms-authentication yet.

**Comment by bricesanchez** [comment]
I don't know why but on production mode on Heroku, i've got this error : 

```shell
2015-04-16T20:52:55.538122+00:00 app[web.1]: Started GET "/" for 208.111.87.10 at 2015-04-16 20:52:55 +0000
2015-04-16T20:52:55.542354+00:00 app[web.1]: Processing by Refinery::PagesController#home as HTML
2015-04-16T20:52:55.836128+00:00 app[web.1]:   Rendered refinery/_content_page.html.erb (277.2ms)
2015-04-16T20:52:55.836823+00:00 app[web.1]: Completed 500 Internal Server Error in 294ms
2015-04-16T20:52:55.836438+00:00 app[web.1]:   Rendered vendor/bundle/ruby/2.2.0/bundler/gems/refinerycms-95d8ec9ea44d/pages/app/views/refinery/pages/home.html.erb within layouts/application (278.2ms)
2015-04-16T20:52:55.839275+00:00 app[web.1]: 
2015-04-16T20:52:55.839278+00:00 app[web.1]: ActionView::Template::Error (undefined method `render_content_page' for #<#<Class:0x007f81584f5408>:0x007f819af4fa60>):
2015-04-16T20:52:55.839280+00:00 app[web.1]:     1: <div class="wrapper">
2015-04-16T20:52:55.839282+00:00 app[web.1]:     2:   <%= render_content_page(@page, {
2015-04-16T20:52:55.839283+00:00 app[web.1]:     3:     :hide_sections => local_assigns[:hide_sections],
2015-04-16T20:52:55.839285+00:00 app[web.1]:     4:     :can_use_fallback => !local_assigns[:show_empty_sections] && !local_assigns[:remove_automatic_sections]
2015-04-16T20:52:55.839286+00:00 app[web.1]:     5:   }) %>
2015-04-16T20:52:55.839289+00:00 app[web.1]:   app/views/refinery/_content_page.html.erb:2:in `_487840ea091a429f76eb5ad01967d4ec'
2015-04-16T20:52:55.839290+00:00 app[web.1]: 
2015-04-16T20:52:55.839291+00:00 app[web.1]: 
2015-04-16T20:52:55.840055+00:00 app[web.1]: cache: [GET /] miss
2015-04-16T20:52:55.975271+00:00 heroku[router]: at=info method=GET path="/" host=j...omm.com request_id=7132c81e-72cb-4f11-9061-8340b8d15a4e fwd="208.111.87.10" dyno=web.1 connect=1ms service=440ms status=500 bytes=1750
2015-04-16T20:52:58.315337+00:00 heroku[router]: at=info method=GET path="/" host=e...e.herokuapp.com request_id=b90b382b-24ed-4993-b8a6-7943d6d75fd3 fwd="54.247.188.179" dyno=web.1 connect=1ms service=11ms status=401 bytes=449
2015-04-16T20:52:58.307839+00:00 app[web.1]: Processing by Refinery::PagesController#home as */*
2015-04-16T20:52:58.308988+00:00 app[web.1]: Completed 401 Unauthorized in 1ms (ActiveRecord: 0.0ms)
2015-04-16T20:52:58.304387+00:00 app[web.1]: Started GET "/" for 54.247.188.179 at 2015-04-16 20:52:58 +0000
2015-04-16T20:52:58.308850+00:00 app[web.1]: Filter chain halted as :authenticate rendered or redirected
2015-04-16T20:52:58.310177+00:00 app[web.1]: cache: [GET /] miss
```

**Comment by parndt** [comment]
This is basically ready to go but needs testing.  I was having issues logging in only.  To try it out, use this branch locally and add the [refinery/refinerycms-authentication-devise](https://github.com/refinery/refinerycms-authentication-devise) extension:

```ruby
gem 'refinerycms-authentication-devise', git: 'https://github.com/refinery/refinerycms-authentication-devise', branch: 'master'
```

You'll need to pull in the latest migrations, too.

**Comment by bricesanchez** [comment]
At short time, i could test without [refinery/refinerycms-authentication-devise](https://github.com/refinery/refinerycms-authentication-devise) but with [bricesanchez/spree-refinery-authentication](https://github.com/bricesanchez/spree-refinery-authentication)

**Comment by parndt** [comment]
That'd be good, because there have been some changes such as needing your user to define `active_plugins` which are the plugins that the user can access that are also activated within the system.  Rather than being string names, they're actual `Refinery::Plugin` objects.

**Comment by simi** [comment]
As I mentioned earlier. It will be better to use `authorize` instead of `authorise`.

File: testing/lib/refinery/testing/feature_macros/authentication.rb [hunk 1]
```diff
@@ -2,13 +2,16 @@ module Refinery
...
     module FeatureMacros
       module Authentication
-        def refinery_login_with(factory)
-          let!(:logged_in_user) { FactoryGirl.create(factory) }
 
-          before do
-            login_as logged_in_user, :scope => :refinery_user
-          end
+        def refinery_login
+          # NO-OP, Overwrite this in an extension.
```
Comment (simi) [inline testing/lib/refinery/testing/feature_macros/authentication.rb]: What about warning here or raising NotImplemented  :question:

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -38,6 +38,13 @@
   config.default_sleep_interval = 0.33
   config.clear_lets_on_failure = true
   config.default_retry_count = 3
+
+  unless ENV['FULL_BACKTRACE']
```
Comment (simi) [inline spec/spec_helper.rb]: :question:

**Comment by simi** [comment]
I think also we can keep `refinery-authentication` instead of moving parts into the core (like NilUser).

Is there any reason why zilch can't be part of refinery (probably part of refinery-authentication)?

**Comment by parndt** [comment]
But a user is a core concept that the entire system uses, no?

File: testing/lib/refinery/testing/feature_macros/authentication.rb (hunk 1 above)
Comment (parndt) [inline testing/lib/refinery/testing/feature_macros/authentication.rb]: But then we'd have to implement it just to run tests in the core system?

File: spec/spec_helper.rb (hunk 2 above)
Comment (parndt) [inline spec/spec_helper.rb]: I got tired of seeing Rails' stack trace for every failed spec. This was much faster/easier to see why tests were failing.

Comment (simi) [inline spec/spec_helper.rb]: I think usage of another rspec formater can do this in more fashionable way.

Comment (parndt) [inline spec/spec_helper.rb]: That'd be nice - can you provide an example of how we can do that? I think you're better at rspec than I am :smile:

**Comment by bricesanchez** [comment]
Usage report : 
- I'm not able to use Refinery in production when eager_load is set to true : https://github.com/bricesanchez/spree-refinery-authentication/issues/7

- I'm not able to load decorators on frontend engine controller decorators, when i do that i've this error : 
```ruby
NoMethodError in Refinery::Jobs::Jobs#index
Showing /Users/bci20/Sites/equ-001-website-2015/app/views/refinery/_content_page.html.slim where line #6 raised:

undefined method `render_content_page' for #<#<Class:0x007fcbf6ab3758>:0x007fcbf6ab2678>
```

**Comment by parndt** [comment]
Okay thanks!

**Comment by bricesanchez** [comment]
@parndt Any idea to fix these issues ?

**Comment by parndt** [comment]
@bricesanchez got a quick sample app?

**Comment by parndt** [comment]
I believe I've fixed the eager loading issues, if @bricesanchez can confirm :smile:

**Comment by bricesanchez** [comment]
i can confirm, eager load issues are fixed. Thanks @parndt :)

**Comment by bricesanchez** [comment]
Awesome work @parndt !

**Comment by robyurkowski** [comment]
This is all @parndt! Amazing work, buddy.

**Pull request merged**